pinot-core/src/main/java/org/apache/pinot/core/data/table/IndexedTable.java

```diff
@@ -37,39 +37,48 @@
   protected final boolean _hasOrderBy;
   protected final TableResizer _tableResizer;
 
-  // The capacity we need to trim to
-  protected final int _capacity;
-  // The capacity with added buffer, in order to collect more records than capacity for better precision
-  protected final int _maxCapacity;
+  // The size we need to trim to
+  protected final int _trimSize;
+  // The size with added buffer, in order to collect more records than capacity for better precision
+  protected final int _trimThresholdForUpsert;
+  protected final int _trimThresholdForFinish;
 
-  protected IndexedTable(DataSchema dataSchema, QueryContext queryContext, int capacity) {
-    super(dataSchema);
+  protected List<Record> _sortedRecords;
 
+  public IndexedTable(DataSchema dataSchema, QueryContext queryContext, int trimSize, int trimThreshold) {
+    super(dataSchema);
     List<ExpressionContext> groupByExpressions = queryContext.getGroupByExpressions();
     assert groupByExpressions != null;
     _numKeyColumns = groupByExpressions.size();
-
     _aggregationFunctions = queryContext.getAggregationFunctions();
-
     List<OrderByExpressionContext> orderByExpressions = queryContext.getOrderByExpressions();
     if (orderByExpressions != null) {
+      // SQL GROUP BY with ORDER BY
+      // trimSize = max (limit N * 5, 5000) (see GroupByUtils.getTableCapacity)
+      // to keep parity with PQL for some use cases with infinitely large group by,
+      // we have different trim thresholds for upsert and finish.
+      // for such cases, trimThresholdForUpsert will be 1_000_000_000 * 2
+      // (exactly same as PQL). this essentially implies there will be no
+      // resizing/trimming during upsert. during finish (called after server combine is over),
+      // we still want to trim the results and trimThreshold is computed in the
+      // same manner as PQL (trimSize * 4)
       _hasOrderBy = true;
       _tableResizer = new TableResizer(dataSchema, queryContext);
-      _capacity = capacity;
-
-      // TODO: tune these numbers and come up with a better formula (github ISSUE-4801)
-      // Based on the capacity and maxCapacity, the resizer will smartly choose to evict/retain recors from the PQ
-      if (capacity
-          <= 100_000) { // Capacity is small, make a very large buffer. Make PQ of records to retain, during resize
-        _maxCapacity = 1_000_000;
-      } else { // Capacity is large, make buffer only slightly bigger. Make PQ of records to evict, during resize
-        _maxCapacity = (int) (capacity * 1.2);
-      }
+      _trimSize = trimSize;
+      _trimThresholdForUpsert = Math.max(trimSize * 4, trimThreshold);
```

Comment:
We should not have this max because the `trimThreshold` is already the higher bound allowed

Reply:
yes, not needed. Removed